enhancements/test-platform/e2e-observer-pods.md

```diff
@@ -0,0 +1,98 @@
+authors:
+  - "@deads2k"
+reviewers:
+approvers:
+  - "@stevek"
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+replaces:
+superseded-by:
+---
+
+# e2e Observer Pods
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+e2e tests have multiple dimensions:
+ 1. which tests are running (parallel, serial, conformance, operator-specific, etc)
+ 2. which platform (gcp, aws, azure, etc)
+ 3. which configuration of that platform (proxy, fips, ovs, etc)
+ 4. which version is running (4.4, 4.5, 4.6, etc)
+ 5. which version change is running (single upgrade, double upgrade, minor upgrade, etc)
+Our individual jobs are an intersection of these dimensions and that intersectionality drove the development of steps
+to help handle the complexity.
+
+There is a kind of CI test observer that wants to run across all of these intersections by leveraging the uniformity of
+OCP across all the different dimensions above.
+It wants to do something like start an observer agent of some kind for every CI cluster ever created and provide o(100M)
+of data back to include in the CI job overall.
+Working with steps would require integrating in multiple dimensions and lead to gaps in coverage.
+
+## Motivation
+
+We need to run tools like
+ 1. e2e monitor - doesn't cover installation
+ 2. resourcewatcher - doesn't cover installation today if we wire it directly into tests.
+ 3. loki log collection (as created by group-b) - hasn't been able to work in upgrades at all.
+ 4. loki as provided by the logging team (future)
+To debug our existing install and upgrade failures.
+
+### Goals
+
+ 1. allow a cluster observing tool to run in the CI cluster (this avoid restarts during upgrades)
+ 2. allow a cluster observer to provide data (including junit) to be collected by the CI job
+
+### Non-Goals
+
+ 1. allow a cluster observer to impact success or failure of a job
+ 2. allow a cluster observer to impact the running test.  This is an observer author failure.
+ 3. provide ANY dimension specific data. If an observer needs this, they need to integrate differently.
+
+## Proposal
+
+### Changes to CI
+This is a sketch of a possible path.
+ 1. Allow a non-dptp-developer to produces a pod template that mounts a `secret/<job>-kubeconfig` that will later contain a kubeconfig.
+ 2. When the CI pod starts today (the one with setup containers and stuff), also create an instance of each pod template,
+    let's say `pod/<job>-resourcewatcher` and an empty `secret/<job>-kubeconfig`.
+ 3. As soon as a kubeconfig is available (these go into a known location in setup container today), write that
```

Comment:
Do we expect the observer process to be _invoked_ at the time the `$KUBECONFIG` is present, or do you expect the process to be running and require that it handle observing the presence or absence of `$KUBECONFIG`?

Reply:
> Do we expect the observer process to be _invoked_ at the time the `$KUBECONFIG` is present, or do you expect the process to be running and require that it handle observing the presence or absence of `$KUBECONFIG`?

that's why #2 was explicit. The e2e observer can be started *before* the kubeconfig is present. In fact, in some clusters, the kubeconfig will never be present because something horrible happened when we tried to provision.

Reply:
> In fact, in some clusters, the kubeconfig will never be present because something horrible happened when we tried to provision.

Do we expect the monitors to be able to provide useful information for these clusters?  How would they discover that information?

I think having the monitors already running and volume-mounting the shared secret (so they can retrieve the kubeconfig) and an artifact volume (so they can report their results, via JUnit, etc.) is sufficient, and as long as they are running by the time the kubeconfig becomes available, we'll be fine.  Launching them in parallel with the first step in the chain (post image builds?) should be fine, although if you wanted to cover the "monitors are very slow to come up" race, you could launch the monitors, wait until they report ready, and then commence with the meat of the test workflow.